Add FRLG

cawtds opened the pull request:
## Description
Adds FRLG to expansion

## Media
-

## Things to note in the release changelog:
-

## Discord contact info
.cawt

hedara90 commented:
Is it in a ready enough state to give comments on?

hedara90 commented:
The `trainers*.h` files are no longer included in the repo by default, they're always built from the `trainers*.party` files and they have their own rule file in `trainer_rules.mk`.
Just adding an `AUTO_GEN_TARGET` for the relevant files should be enough.

ShinyDragonHunter commented:
If you're doing something like this, it might be worth updating Emerald's data (using Cut on Grass etc) to match FRLG or adapting FRLG's data to match Emerald as otherwise it's gonna be a nightmare for maintaining. FRLG also handled forced movement actions differently (uses one array as opposed to two in Emerald), you could think about what's better in either game and port.

I've also been able to set things up like arrays for animated TVs and escalators which check for tileset when animating. I can pass along changes if it'd be something in scope

hedara90 commented on data/battle_anim_scripts.s:
```diff
@@ -14,6 +14,58 @@
+gBattleAnimGeneral_MonScared::
+	createvisualtask AnimTask_SetAttackerTargetLeftPos, 2, 1
+	waitforvisualfinish
+	loadspritegfx ANIM_TAG_SWEAT_BEAD
```
Just a note here for the future.
#8497 removes `loadspritegfx` and integrates all asset loading into the commands that use them.

hedara90 commented on data/battle_anim_scripts.s:
```diff
@@ -14,6 +14,58 @@
+	waitforvisualfinish
+	clearmonbg_static ANIM_ATTACKER
+	delay 1
+	loadspritegfx ANIM_TAG_SWEAT_BEAD
```
Same as previous comment, but `loadspritegfx` also has an implicit `delay 1` built into it, I don't know if it's relevant here.

hedara90 commented:
Important stuff:
- Marowak Ghost doesn't have moves
- Coins picked up from the floor in Game Corner says ????????
 
Graphical stuff:
- Pokedex count in save screen uses only mons in the Hoenn dex for the count
- Badges displayed in the trainer card are not aligned with the boxes
- Marowak Ghost is shaded (I don't know if it should be or not)

Text stuff:
- Transferring a caught mon to the PC says "Lanette's" instead of "Bill's"

Places that are "outdoors" but not marked as such (for purposes of DNS):
- SS Anne docks
- SS Anne outdoor deck
- Roof of Celadon Condominiums
- Roof of Celadon Department store

Not necessary for initial merge:
- Map previews are currently missing

cawtds commented:
> Important stuff:
> 
>     * Marowak Ghost doesn't have moves
> 
>     * Coins picked up from the floor in Game Corner says ????????
> 
> 
> Graphical stuff:
> 
>     * Pokedex count in save screen uses only mons in the Hoenn dex for the count
> 
>     * Badges displayed in the trainer card are not aligned with the boxes
> 
>     * Marowak Ghost is shaded (I don't know if it should be or not)
> 
> 
> Text stuff:
> 
>     * Transferring a caught mon to the PC says "Lanette's" instead of "Bill's", also using "Lanette" when accessing the PC.
> 

✔️  Fixed


> Places that are "outdoors" but not marked as such (for purposes of DNS):
> 
>     * SS Anne docks
> 
>     * SS Anne outdoor deck
> 
>     * Roof of Celadon Condominiums
> 
>     * Roof of Celadon Department store

Not sure on all the side effects this would have, so I'd rather not change it in this PR

> FRLG features currently not working
> 
>     * Map previews
> 
>     * Fame checker

Both new features that do not affect FRLG gameplay. So imo new PRs

hedara90 commented on include/battle.h:
```diff
@@ -1193,4 +1194,9 @@ static inline void SetHealAmount(u32 battler, s32 value)
     gBattleStruct->passiveHpUpdate[battler] = -1 * value;
 }
 
+static inline bool32 IsGhostBattleWithoutScope()
```
```suggestion
static inline bool32 IsGhostBattleWithoutScope(void)
```

hedara90 commented on include/field_control_avatar.h:
```diff
@@ -34,6 +34,8 @@ const u8 *GetInteractedLinkPlayerScript(struct MapPosition *position, u8 metatil
 const u8 *GetCoordEventScriptAtMapPosition(struct MapPosition *position);
 void ClearPoisonStepCounter(void);
 void CancelSignPostMessageBox(struct FieldInput *input);
+void HandleBoulderFallThroughHole(struct ObjectEvent *);
```
```suggestion
void HandleBoulderFallThroughHole(struct ObjectEvent *object);
```

cawtds replied:
✔️

hedara90 commented on include/map_preview_screen.h:
```diff
@@ -0,0 +1,64 @@
+void MapPreview_LoadGfx(u8 mapsec);
+bool32 MapPreview_IsGfxLoadFinished(void);
+void MapPreview_Unload(s32 windowId);
+void MapPreview_StartForestTransition(u8 mapsec);
```
The `mapsec` arguments here should probably the `mapsec_u8_t` types.

cawtds replied:
✔️

hedara90 commented on include/title_screen.h:
```diff
@@ -4,5 +4,7 @@
 
 void CB2_InitTitleScreen(void);
+void CB2_InitTitleScreenFrlg(void);
+
```
```suggestion
```

cawtds replied:
✔️

hedara90 commented on src/battle_controller_opponent.c:
Making a note that this file has potential for cleanup post-merge

hedara90 commented on src/battle_controller_player.c:
```diff
@@ -1885,7 +1893,7 @@ static enum TrainerPicID PlayerGetTrainerBackPicId(void)
         trainerPicId = LinkPlayerGetTrainerPicId(GetMultiplayerId());
     else
-        trainerPicId = gSaveBlock2Ptr->playerGender + TRAINER_PIC_BACK_BRENDAN;
+        trainerPicId = gSaveBlock2Ptr->playerGender ? TRAINER_BACK_PIC_PLAYER_FEMALE : TRAINER_BACK_PIC_PLAYER_MALE;
```
```suggestion
        trainerPicId = gSaveBlock2Ptr->playerGender == FEMALE ? TRAINER_BACK_PIC_PLAYER_FEMALE : TRAINER_BACK_PIC_PLAYER_MALE;
```
Non-boolean variable treated as boolean.

cawtds replied:
✔️

hedara90 commented on src/battle_controller_safari.c:
```diff
@@ -275,7 +275,7 @@ void SafariBufferExecCompleted(u32 battler)
 static void SafariHandleDrawTrainerPic(u32 battler)
 {
-    enum TrainerPicID trainerPicId = gSaveBlock2Ptr->playerGender + TRAINER_PIC_BACK_BRENDAN;
+    enum TrainerPicID trainerPicId = gSaveBlock2Ptr->playerGender == MALE ? TRAINER_BACK_PIC_PLAYER_MALE : TRAINER_BACK_PIC_PLAYER_FEMALE;
```
This might be good to change to match the order the previous handling for this is done in where female is the `TRUE` case

hedara90 commented on src/battle_script_commands.c:
```diff
@@ -1119,6 +1119,23 @@ static void AccuracyCheck(bool32 recalcDragonDarts, const u8 *nextInstr, const u
+        ((gBattleTypeFlags & BATTLE_TYPE_FIRST_BATTLE && (!BtlCtrl_OakOldMan_TestState2Flag(1) || !BtlCtrl_OakOldMan_TestState2Flag(2))
+        && gMovesInfo[gCurrentMove].power != 0 && GetBattlerSide(gBattlerAttacker) == B_SIDE_PLAYER)
+        || (gBattleTypeFlags & BATTLE_TYPE_POKEDUDE)))
+    {    if (gBattleStruct->moveResultFlags[gBattlerTarget] & MOVE_RESULT_MISSED)
```
Can this `if` statement be cleaned up a bit? It's hard to track how the conditions are grouped here.
```C
    if (IS_FRLG
     && ((gBattleTypeFlags & BATTLE_TYPE_FIRST_BATTLE
       && (!BtlCtrl_OakOldMan_TestState2Flag(1) || !BtlCtrl_OakOldMan_TestState2Flag(2))
       && gMovesInfo[gCurrentMove].power != 0
       && GetBattlerSide(gBattlerAttacker) == B_SIDE_PLAYER)
      || (gBattleTypeFlags & BATTLE_TYPE_POKEDUDE)))
    {
        if (gBattleStruct->moveResultFlags[gBattlerTarget] & MOVE_RESULT_MISSED)
```
Should work, I don't dare to make the suggestion in the Github UI

hedara90 commented on src/credits_frlg.c:
```diff
@@ -0,0 +1,1558 @@
+};
+
+EWRAM_DATA bool8 gHasHallOfFameRecords = 0;
+static EWRAM_DATA struct CreditsResources * sCreditsMgr = NULL;
```
```suggestion
static EWRAM_DATA struct CreditsResources *sCreditsMgr = NULL;
```

hedara90 commented on src/credits_frlg.c:
Clean up formatting in this file.
(Can be done after merge)

hedara90 commented on src/data/object_events/object_event_graphics_info.h:
Clean up formatting in this file. Also applies to existing `struct`s in this file.
(Can be done after merge)

hedara90 commented on src/field_specials.c:
Formatting should be cleaned up in this file.
(Can happen after merge)

hedara90 commented on src/intro_frlg.c:
Clean up `EWRAM` variables and `struct`s.
Can be done after merge.

hedara90 commented on src/map_preview_screen.c:
Clean up formatting.
(Can be done after merge)

hedara90 commented on src/menu.c:
```diff
@@ -1959,3 +1964,69 @@ void HBlankCB_DoublePopupWindow(void)
+
+static void Task_SmoothBlendLayers(u8 taskId)
+{
+    s16 * data = gTasks[taskId].data;
```
```suggestion
    s16 *data = gTasks[taskId].data;
```

hedara90 commented on src/script_menu.c:
```diff
@@ -748,7 +749,11 @@ static void CreatePCMultichoice(void)
         AddTextPrinterParameterized(windowId, FONT_NORMAL, gText_LanettesPC, x, 1, TEXT_SKIP_DRAW, NULL);
+#endif
     else
         AddTextPrinterParameterized(windowId, FONT_NORMAL, gText_SomeonesPC, x, 1, TEXT_SKIP_DRAW, NULL);
```
```suggestion
    if (FlagGet(FLAG_SYS_PC_LANETTE))
    {
        if (IS_FRLG)
            AddTextPrinterParameterized(windowId, FONT_NORMAL, gText_BillsPc, x, 1, TEXT_SKIP_DRAW, NULL);
        else
            AddTextPrinterParameterized(windowId, FONT_NORMAL, gText_LanettesPC, x, 1, TEXT_SKIP_DRAW, NULL);
    }
    else
    {
        AddTextPrinterParameterized(windowId, FONT_NORMAL, gText_SomeonesPC, x, 1, TEXT_SKIP_DRAW, NULL);
    }
```

hedara90 commented:
Conflicts

cawtds commented on include/battle_gfx_sfx_util.h:
```diff
@@ -12,6 +12,7 @@ bool8 TryHandleLaunchBattleTableAnimation(u8 activeBattlerId, u8 attacker, u8 ta
 void InitAndLaunchSpecialAnimation(u8 activeBattlerId, u8 attacker, u8 target, u8 tableId);
 bool8 IsBattleSEPlaying(u8 battler);
 void BattleLoadMonSpriteGfx(struct Pokemon *mon, u32 battler);
+void DecompressGhostFrontPic(u32 battlerId);
```
✔️

cawtds commented:
> Emotes when using a game corner slot machine are broken. 

fixed
> Trainer Sharon on Kindle Road has a broken script. Trainer Yuji in Viridian Gym has a broken script. 

also broken in emerald, unrelated to this PR
> Hidden Leppa Berry on Route 23 is doesn't work.

cause not clear yet
> Zigzagoon battle in Emerald is triggering the Oak Party tutorial.

fixed
